### DISCUSSION TITLE
Is it possible to compile tree-sitter grammars with wasm-pack?

### DISCUSSION THREAD

**Discussion by ahelwer** [tree-sitter/tree-sitter#2010] [Q&A] [answered]
I wrote [a fun utility](https://github.com/tlaplus-community/tlauc) using the rust bindings for the [TLA+ tree-sitter grammar](https://github.com/tlaplus-community/tree-sitter-tlaplus) and would like to whip up a web demo so people don't have to download it to try it out. To that end, I tried following instructions to use `wasm-pack` to build my rust utility. Unfortunately it fails when building tree-sitter-tlaplus, with:
```
  running: "clang" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=wasm32-unknown-unknown" "-I" "src" "-Wall" "
-Wextra" "-o" "/home/ahelwer/src/tlaplus/tree-sitter-tlaplus/target/wasm32-unknown-unknown/release/build/tree-sitter-tlaplus-90
d9bfea3134c6dd/out/src/parser.o" "-c" "src/parser.c"
  cargo:warning=In file included from src/parser.c:1:
  cargo:warning=src/tree_sitter/parser.h:10:10: fatal error: 'stdlib.h' file not found
  cargo:warning=#include <stdlib.h>
  cargo:warning=         ^~~~~~~~~~
  cargo:warning=1 error generated.
  exit status: 1
```
From poking around it seems like I need to pass the libc path to clang in the tree-sitter-tlaplus build.rs or something. Emscripten takes care of this with a built-in libc. So do I need to find some webassembly port of libc and use that, then? Is it statically included in the final product, or will I have to ensure the webassembly runtime environment has access to the same libc port? Sorry, I'm a bit of a webassembly newbie.

Alternatively, is there some way of getting wasm-pack to link against the webassembly version of tree-sitter-tlaplus built using emscripten, since that's how we usually build it around here?

**Comment by liamrosenfeld** [comment]
You can provide a minimal c sysroot (like [this](https://github.com/liamrosenfeld/tree-sitter-python/tree/2c082b5a0a1d1640c9a01b852e95e980a362fb54/bindings/rust/wasm-sysroot)) and then add it to your `build.rs` (like [this](https://github.com/liamrosenfeld/tree-sitter-python/blob/2c082b5a0a1d1640c9a01b852e95e980a362fb54/bindings/rust/build.rs))

**Reply by StringKe** [reply]
For example, https://github.com/tree-sitter/tree-sitter-typescript does not provide similar support.

Do I need to make a new version of tree-sitter-typescript?

**Comment by ObserverOfTime** [comment, marked as answer]
Fixed in #4820
